java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c

```diff
@@ -350,6 +350,85 @@ JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addFre
+  (void) jobj; // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtSessionOptions* options = (OrtSessionOptions*) optionsHandle;
+
+  size_t arrLength = (*jniEnv)->GetArrayLength(jniEnv, handlesArray);
+
+  const char** names = allocarray(arrLength, sizeof(char*));
+  if (names == NULL) {
+    // Nothing to cleanup, return and throw exception
+    throwOrtException(jniEnv, 1, "Not enough memory");
+    return;
+  }
+  jobject* javaNameStrings = allocarray(arrLength, sizeof(jobject));
+  if (javaNameStrings == NULL) {
+    goto cleanup_names;
+  }
+  const OrtValue** initializers = allocarray(arrLength, sizeof(OrtValue*));
+  if (initializers == NULL) {
+    goto cleanup_java_input_strings;
+  }
+
+  // Extract a C array of longs which are pointers to the input tensors.
+  // The Java-side objects store native pointers as 64-bit longs, and on 32-bit systems
+  // we cannot cast the long array to a pointer array as they are different sizes,
+  // so we copy the longs applying the appropriate cast.
+  jlong* initializersArr = (*jniEnv)->GetLongArrayElements(jniEnv, handlesArray, NULL);
+
+  for (size_t i = 0; i < arrLength; i++) {
+    // Extract the string chars and cast the tensor
+    javaNameStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv, namesArray, (jint) i);
+    names[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaNameStrings[i], NULL);
+    initializers[i] = (const OrtValue*) initializersArr[i];
+  }
+
+  checkOrtStatus(jniEnv,api,api->AddExternalInitializers(options,names,initializers,arrLength));
+
+  // Release the java array copy of pointers to the tensors.
+  (*jniEnv)->ReleaseLongArrayElements(jniEnv, handlesArray, initializersArr, JNI_ABORT);
+  free(initializers);
+cleanup_java_input_strings:
+  // Release the Java strings
+  for (size_t i = 0; i < arrLength; i++) {
+    (*jniEnv)->ReleaseStringUTFChars(jniEnv, javaNameStrings[i], names[i]);
+  }
+  free(javaNameStrings);
+cleanup_names:
+  free(names);
+}
+
+/*
+ * Class:     ai_onnxruntime_OrtSession_SessionOptions
+ * Method:    addInitializer
+ * Signature: (JJLjava/lang/String;J)V
+ */
+JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addInitializer
+    (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong optionsHandle, jstring name, jlong tensorHandle) {
+  (void) jobj; // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtSessionOptions* options = (OrtSessionOptions*) optionsHandle;
+
+  // Extract the string chars
+  const char* cName = (*jniEnv)->GetStringUTFChars(jniEnv, name, NULL);
+
+  // Cast the onnx value
+  const OrtValue* tensor = (const OrtValue*) tensorHandle;
+
+  checkOrtStatus(jniEnv,api,api->AddInitializer(options,cName,tensor));
```

Comment:
This function throws on errors? Won't that skip over the ReleaseStringUTFChars?

Reply:
Same comment as above, the checkOrtStatus method throws a Java exception, but the JNI code completes before it is propagated.